### DISCUSSION TITLE
v0.8 IfcCurveSegment geometry mapping architecture problems

### DISCUSSION THREAD

**Discussion by RickBrice** [IfcOpenShell/IfcOpenShell#3934] [General]
@aothms @civilx64 
It looks like there is an architecture problem with v0.8 for mapping IfcCurveSegment for IfcCompositeCurve, IfcGradientCurve, and IfcSegmentedReferenceCurve.

As I understand the spec, and supported by discussions (https://github.com/buildingSMART/IFC4.3.x-development/issues/702), the IfcCurveSegment.Placement of one segment defines the end point of the previous IfcCurveSegment, which leads to the need for a zero-length segment at the end of the curve. Numeric errors in the generated geometry can be corrected knowing the expected end point.

For IfcSegmentReferenceCurve, the cant vertical deviation and rate of rotation is defined by the IfcCurveSegment.Placement for each consecutive segment pair. For cant 
> If no deviation in the position of the curve segment to the base curve axis is specified, the axes (Axis and RefDirection) directions of [IfcAxis2Placement](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/lexical/IfcAxis2Placement.htm) are **interpolated between the initial curve segment placement and the placement of the subsequent curve segment**.

To support this, I think that `mapping::map_impl(const IfcSchema::IfcCurveSegment* inst)` needs to have two parameters, `mapping::map_impl(const IfcSchema::IfcCurveSegment* inst,const IfcSchema::IfcCurveSegment* next_inst)`

The `mapping::map_impl` method for IfcCompositeCurve, IfcGradientCurve, and (under construction) IfcSegmentedReferenceCurve all have a loop that traverses the individual IfcCurveSegment entities composing the curve

https://github.com/IfcOpenShell/IfcOpenShell/blob/2a53911c52bc6857ef0afd9ea67e460cda51d73c/src/ifcgeom/mapping/IfcGradientCurve.cpp#L32-L47

This loop may need to be changed to something like this pseudo-code,

~~~
auto iter = segments->begin();
auto next = std::next(iter);
auto end = std::prev(segments->end());
for( ; iter != end; iter++, next++)
{
   auto& this_segment(*iter);
   auto& next_segment(*next);
   auto crv = map(this_segment->as<IfcSchema::IfcCurveSegment>(),next_segment->as<IfcSchema::IfcCurveSegment>());
...
}
~~~

However, `map` is defined as a pure-virtual method in `abstract_mapping`.

A solution could be to add a mapping function declaration to abstract_mapping that takes two parameters (maybe call it map_adjacement or map_related or just overload map).

`virtual ifcopenshell::geometry::taxonomy::ptr map(const IfcUtil::IfcBaseInterface*,const IfcUtil::IfcBaseInterface*) = 0;`

Then implement in `mapping` as

~~~
taxonomy::ptr mapping::map(const IfcBaseInterface* inst,const IfcBaseInterface* next_inst) {
	auto iden = inst->as<IfcUtil::IfcBaseClass>()->identity();
	auto it = cache_.find(iden);
	if (it != cache_.end()) {
		return it->second;
	}
	taxonomy::ptr item = nullptr;

	// @todo we should check whether there is a notice performance impact on the large sequence
	// of if-statements and whether a switch on e.g inst->declaration()->index_in_schema()
	// isn't more efficient (which would disable inheritance though).

#include "bind_pair_convert_impl.i"  // << NOTE CHANGE HERE

	if (item) {
		cache_.insert({ iden, item });
	}
	else {
		Logger::Message(Logger::LOG_ERROR, "No operation defined for:", inst);
	}
	return item;
}
~~~

bind_pair_convert_impl.i would be similar to bind_convert_impl.i except that `item = map_impl(inst->as<IfcSchema::T>());` would become `item = map_impl(inst->as<IfcSchema::T>(),next_inst->as<IfcSchema::T>());`

A new `bind_pair_convert_decl.i` is then needed

~~~
#ifdef BIND_PAIR
#undef BIND_PAIR
#endif

#define BIND_PAIR(T) ifcopenshell::geometry::taxonomy::ptr map_impl(const IfcSchema::T*,const IfcSchema::T*);

#include "mapping_pair.i"
~~~

And a new `mapping_pair.i` is needed, removing `BIND(IfcCurveSegment);` from mapping.i

~~~
#ifdef SCHEMA_HAS_IfcCurveSegment
BIND_PAIR(IfcCurveSegment);
#endif
~~~

What do you think?

**Comment by aothms** [comment]
Thanks for the detailed write up.

One other thing we can do is go backwards over the unnamed implicit inverse associated with the Segments attribute. Lookup the index in those Segments from the 'parent' and keep note of the segment following that. See code sketch below:

```C++
taxonomy::ptr mapping::map_impl(const IfcSchema::IfcCurveSegment* inst) {
	// @todo: rb figure out what to do with the zero length segments at the end of compound curves

	IfcSchema::IfcCurveSegment* next = nullptr;
	auto compcurves = inst->data().getInverse(&IfcSchema::IfcCompositeCurve::Class(), 0);
	if (compcurves && compcurves->size() == 1) {
		auto segments = (*compcurves->begin())->as<IfcSchema::IfcCompositeCurve>()->Segments();
		bool emit_next = false;
		for (auto& s : *segments) {
			if (emit_next) {
				next = s->as<IfcSchema::IfcCurveSegment>();
				break;
			}
			if (s == inst) {
				emit_next = true;
			}
		}
	}
```

In case there are multiple composite curves using the same segments that wouldn't work, because we can never recover the context in which the segment is being processed. This, as opposed to the approach you are presenting. One downside of the approach you are presenting is that it will likely become harder again to get the representation of a single IfcAlignmentSegment represented by an IfcCurveSegment directly, because in this case the call stack does not include evaluating the composite.

So I would propose this inverse lookup approach for now, as is that it is a little bit less intrusive and that there might be other cases where some sort of context is passed to the mapping function. I was thinking about using something like bgl named_parameters for that https://www.boost.org/doc/libs/1_83_0/libs/graph/doc/bgl_named_params.html

**Reply by RickBrice** [reply]
@aothms Your response is exactly why I wanted to ask before proceeding - thanks!

> auto compcurves = inst->data().getInverse(&IfcSchema::IfcCompositeCurve::Class(), 0);
I've seen lines of code like this before but never had reason to dig into what it actually means.

It seems inefficient to loop over all the segments in the composite every time to find the next segment, but I don't think there will be too many segments per curve in practice.

I'll proceed with your approach and let you know if I run into any problems.
